generator.go

```diff
@@ -204,52 +205,58 @@ func (g *Generator) getGenerateInputs() buildpack.GenerateInputs {
 	}
 }
 
-func (g *Generator) copyDockerfiles(dockerfiles []buildpack.DockerfileInfo, newBaseIdx int) error {
-	for currentIdx, dockerfile := range dockerfiles {
+func (g *Generator) copyDockerfiles(dockerfiles []buildpack.DockerfileInfo) error {
+	for _, dockerfile := range dockerfiles {
 		targetDir := filepath.Join(g.GeneratedDir, dockerfile.Kind, launch.EscapeID(dockerfile.ExtensionID))
 		var targetPath = filepath.Join(targetDir, "Dockerfile")
-		if dockerfile.Kind == buildpack.DockerfileKindRun && currentIdx < newBaseIdx {
+		if dockerfile.Kind == buildpack.DockerfileKindRun && dockerfile.Ignore {
 			targetPath += ".ignore"
 		}
 		if err := os.MkdirAll(targetDir, os.ModePerm); err != nil {
 			return err
 		}
 		g.Logger.Debugf("Copying %s to %s", dockerfile.Path, targetPath)
 		if err := fsutil.Copy(dockerfile.Path, targetPath); err != nil {
-			return err
+			return fmt.Errorf("failed to copy Dockerfile at %s: %w", dockerfile.Path, err)
 		}
 		// check for extend-config.toml and if found, copy
 		extendConfigPath := filepath.Join(filepath.Dir(dockerfile.Path), "extend-config.toml")
 		if err := fsutil.Copy(extendConfigPath, filepath.Join(targetDir, "extend-config.toml")); err != nil {
 			if !os.IsNotExist(err) {
-				return err
+				return fmt.Errorf("failed to copy extend config at %s: %w", extendConfigPath, err)
 			}
 		}
 	}
 	return nil
 }
 
-func (g *Generator) checkNewRunImage(dockerfiles []buildpack.DockerfileInfo) (newBase string, newBaseIdx int, extend bool) {
-	// There may be extensions that contribute only a build.Dockerfile; work backward through extensions until we find
-	// a run.Dockerfile.
+func (g *Generator) runImageFrom(dockerfiles []buildpack.DockerfileInfo) (newBase string, extend bool) {
+	var ignoreNext bool
 	for i := len(dockerfiles) - 1; i >= 0; i-- {
+		// There may be extensions that contribute only a build.Dockerfile;
+		// work backward through extensions until we find a run.Dockerfile.
 		if dockerfiles[i].Kind != buildpack.DockerfileKindRun {
 			continue
 		}
-		if dockerfiles[i].NewBase != "" {
-			newBase = dockerfiles[i].NewBase
-			newBaseIdx = i
-			g.Logger.Debugf("Found a run.Dockerfile configuring image '%s' from extension with id '%s'", newBase, dockerfiles[i].ExtensionID)
-			break
+		if ignoreNext {
+			// If a run.Dockerfile following this one (in the build, not in the loop) switches the run image,
+			// we can ignore this run.Dockerfile as it has no effect.
```

Comment:
```suggestion
			// we can ignore this run.Dockerfile as it has no effect.
			// We set Ignore to true so that when the Dockerfiles are copied to the "generated" directory,
			// we'll add the suffix `.ignore` so that the extender won't try to apply them.			
```